### PR TITLE
Close #26072: Don't show sync menu item until account manager is available

### LINKED ISSUE #26072: Sign in to sync toolbar menu item, indicates I'm logout when I'm not
### Steps to reproduce

Prerequisites authenticate with your Firefox account. 

1. Kill Nightly.
2. Open Nightly.
3. Tab the 3 dotted menu. 

### Expected behaviour

The menu item should indicate that I'm authenticated 

### Actual behaviour

The menu item indicates I'm not authenticated 

### Device name

_No response_

### Android version

Android 12

### Firefox release type

Firefox Nightly

### Firefox version

104.0a1

### Device logs

_No response_

### Additional information

_No response_

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21040)

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #26072

### CODE REVIEW

**Comment by MozillaNoah** [comment]
I can unfortunately still reproduce the bug :(. 

It also looks like this bug is being tracked in: 
- https://github.com/mozilla-mobile/fenix/issues/25486
- https://github.com/mozilla-mobile/fenix/issues/26050

**Comment by rocketsroger** [comment]
> I can unfortunately still reproduce the bug :(
> 
> It also looks like this bug is being tracked with some discussion in:
> 
>     * [App menu shows "Sign in to sync" menu item until I tap it, even though I'm already signed in to sync #25486](https://github.com/mozilla-mobile/fenix/issues/25486)
> 
>     * [[Bug]: FxA account menu item doesn't always reflect latest account state #26050](https://github.com/mozilla-mobile/fenix/issues/26050)

This can only be reproduced when it's a fresh start with no tabs opened.  Signed into account, close all tabs, swipe away Fenix (Force stop)  then start Fenix again.  Once started (should be showing home) tap on menu.

**Comment by MozillaNoah** [comment]
> > I can unfortunately still reproduce the bug :(
> > It also looks like this bug is being tracked with some discussion in:
> > ```
> > * [App menu shows "Sign in to sync" menu item until I tap it, even though I'm already signed in to sync #25486](https://github.com/mozilla-mobile/fenix/issues/25486)
> > 
> > * [[Bug]: FxA account menu item doesn't always reflect latest account state #26050](https://github.com/mozilla-mobile/fenix/issues/26050)
> > ```
> 
> This can only be reproduced when it's a fresh start with no tabs opened. Signed into account, close all tabs, swipe away Fenix (Force stop) then start Fenix again. Once started (should be showing home) tap on menu.

For some clarification, are you meaning that this fix is _just_ for the bug while on Home and not any first app open?

**Comment by rocketsroger** [comment]
> For some clarification, are you meaning that this fix is _just_ for the bug while on Home and not any first app open?

Yes, `HomeMenu` is the one we use in Home.  When opening a session we use the `DefaultToolbarMenu`.

**Comment by rocketsroger** [comment]
Just thought of another use case that can easily produce this issue.  Restart Fenix in private mode will also consistently reproduce this problem since we don't restore session in private mode so we always shows home.

**Comment by Mugurell** [comment]
This seems to just move the issue elsewhere.
If we delay constructing the menu until some later time then users won't be able to interact with it until that later time. And this would affect all users, logged in or not.

https://user-images.githubusercontent.com/11428869/185069343-ce24b59c-fccc-4114-b9ef-653e3f086986.mp4

**Comment by rocketsroger** [comment]
> This seems to just move the issue elsewhere. If we delay constructing the menu until some later time then users won't be able to interact with it until that later time. There is also a visible difference. And this would affect all users, logged in or not.

Hmm, from my testing I never seen that.  Code wise I knew that could happen but account manager was always ready before we finish drawing everything.  We could update the home menu after but it's weird to create the menu twice when we usually just need it to be created when we have updated information.

**Comment by rocketsroger** [comment]
From my testing when the user is not signed in the account manager is ready much faster as well.

**Comment by rocketsroger** [comment]
Here's what I got on my Pixel 3 XL with no account signed in

**Comment by rocketsroger** [comment]
I'll see if I can find a better solution today.

**Comment by rocketsroger** [comment]
@Mugurell After some thoughts, to fix the issue without any change in behaviour, I've just added an extra build after the account is ready.  If the account manager is already ready then we don't do the extra rebuild.  Not sure if this is a good solution.

**Comment by Mugurell** [comment]
What I'm now seeing is that the menu initially opens with "Sign in to sync" and when the account manager is available the menu is automatically closed.

Result on a OnePlus7 with Android 11:

https://user-images.githubusercontent.com/11428869/185310214-20d4909b-3de6-46c5-8017-aac886eef63a.mp4

Given that the menu width is dynamic based on the widest element tinkering with it to switch to showing the email address when available and replacing "Sign in to sync" would probably lead to other issues: either the menu will keep the same width and the email will be ellipsized, either the menu will be invalidated and it would get all of a sudden wider to allow for showing the entire email.

To me this seems like a scenario that should be resolved by UX by maybe introducing another string beside "Sign in to sync" to express that connecting to FXA is currently in progress and the app does not know at that time whether the user is authenticated or not - similar to how in other cases when we have to fetch data we have a way of indicating progress for the meantime.
Saw a discussion about this and the UX being tagged in https://github.com/mozilla-mobile/fenix/issues/25486#issuecomment-1206411322.

**Comment by rocketsroger** [comment]
> What I'm now seeing is that the menu initially opens with "Sign in to sync" and when the account manager is available the menu is automatically closed.

I understand the concern.  But I think this issue applies to users who are signed into their account and tap on the menu right after a fresh start.  I'll create an issue to track removing this when we have a better solution.  But I think this is a small price to pay so the user is frequently confused that they've been signed out.

**Comment by rocketsroger** [comment]
Created the issue to track removing the workaround https://github.com/mozilla-mobile/fenix/issues/26526.  Updated the patch to document it.

**Comment by gabrielluong** [comment]
Can you check if this is related https://github.com/mozilla-mobile/fenix/pull/21488?

**Comment by rocketsroger** [comment]
> Can you check if this is related #21488?

I'll look into it.  Thanks

**Comment by rocketsroger** [comment]
@Mugurell, I took the suggestion from @mcarare to hide the menu item until it's ready.  This seems like the best middle ground solution.  Please let me know if this is acceptable.  Thanks

File: app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt [hunk 1]
```diff
@@ -96,15 +96,26 @@ class HomeMenu(
...
-        R.drawable.ic_synced_tabs,
-        primaryTextColor
-    ) {
-        onItemTapped.invoke(Item.SyncAccount(accountManager.accountState))
+        return when (syncItemTitle) {
+            null -> null
+            else -> {
+                BrowserMenuImageText(
+                    syncItemTitle,
+                    R.drawable.ic_signed_out,
```
Comment (rocketsroger) [inline app/src/main/java/org/mozilla/fenix/home/HomeMenu.kt:112]: I noticed that the icon doesn't match the toolbar menu. I've updated the icon to ic_signed_out here but would require confirmation from UX.

**Comment by rocketsroger** [comment]
account manager not ready (left) and account manager ready (right)
Screenshot_20220818_170150

**Comment by Mugurell** [comment]
> @Mugurell, I took the suggestion from @mcarare to hide the menu item until it's ready. This seems like the best middle ground solution. Please let me know if this is acceptable. Thanks

Agree that this seems like the best middle ground. Don't see it causing other issues. 👍

**Review by Mugurell — APPROVED** [review]

**Comment by rocketsroger** [comment]
Thanks! I'll land this after the cut.

**Pull request merged**